### PR TITLE
DFAST-219: UCrit is not correct in GUI after new selection of reach

### PR DESCRIPTION
**Pull request by ralph-peelen-deltares**
- set default expected ucrit value in ucrit label for the selected reach and branch
- make caching possible for ucrit during application runtime usage
- take in account ucrit value loaded from app configuration, cache this too

### CODE REVIEW

File: dfastmi/gui/dialog_view.py [hunk 1]
```diff
@@ -207,6 +210,20 @@ def _update_sedimentation_length(self, slength: str):
...
 
+    def _update_ucritical(self, ucrit: float, default: float):
+        """
+        Update the GUI components when the critical (minimum) velocity [m/s] for sediment transport changes.
+
+        Args:
+            ucrit: The critical (minimum) velocity [m/s] for sediment transport.
+            default: The default critical (minimum) velocity [m/s] for sediment transport.
+        """
+        # Update the threshold discharge in the GUI
```
Comment (MRVermeulenDeltares) [inline dfastmi/gui/dialog_view.py:221]: Does this comment add value?

File: dfastmi/gui/dialog_view_model.py [hunk 2]
```diff
@@ -309,8 +338,13 @@ def _initialize_ucritical(self):
         """
         Initialize the critical velocity.
         """
-        if self.model.ucritical < self.current_reach.ucritical:
-            self.model.ucritical = self.current_reach.ucritical
+        self._ucritical = 0.0
+        if (self.current_branch, self.current_reach) in self._ucrit_cache:
```
Comment (MRVermeulenDeltares) [inline dfastmi/gui/dialog_view_model.py:342]: maybe put this in a descriptive method.

File: dfastmi/io/AReach.py [hunk 3]
```diff
@@ -55,6 +55,17 @@ def __init__(self, reach_name: str = "Reach", reach_config_key_index: int = 1):
...
         self._config_key_index = reach_config_key_index
 
+    def __hash__(self):
+        return hash((self._name, self._config_key_index))
+
+    def __eq__(self, other):
+        if not isinstance(other, AReach):
+            return False
+        return (
+            self._name == other._name
```
Comment (MRVermeulenDeltares) [inline dfastmi/io/AReach.py:65]: The == should only check on name and index, not check on other values of the Areach object?
e.g. normal_width, ucritical, qstagnant, parent_branch

**Review by MRVermeulenDeltares — APPROVED** [review]
Looks ok, please have a look at my comments before moving to test.

File: dfastmi/gui/dialog_view.py (hunk 1 above)
Comment (ralph-peelen-deltares) [reply dfastmi/gui/dialog_view.py:221]: will keep to stay equal to similar functions :
- _update_branch
- _update_reach
- _update_sedimentation_length
- _update_qthreshold

**Pull request merged**